core/src/apps/webauthn/__init__.py

```diff
@@ -749,3 +1230,413 @@ def msg_error(cid: int, code: int) -> Cmd:
+        or not isinstance(user_verification, bool)
+    ):
+        return cbor_error(req.cid, _ERR_INVALID_CBOR)
+
+    if resident_key and not _ALLOW_RESIDENT_CREDENTIALS:
+        return cbor_error(req.cid, _ERR_UNSUPPORTED_OPTION)
+
+    if user_verification and not config.has_pin():
+        # User verification requested, but PIN is not enabled.
+        state_set = dialog_mgr.set_state(Fido2ConfirmNoPin(req.cid, dialog_mgr.iface))
+    else:
+        # Ask user to confirm registration.
+        state_set = dialog_mgr.set_state(
+            Fido2ConfirmMakeCredential(
+                req.cid,
+                dialog_mgr.iface,
+                client_data_hash,
+                cred,
+                resident_key,
+                user_verification,
+            )
+        )
+
+    if not state_set:
+        return cmd_error(req.cid, _ERR_CHANNEL_BUSY)
+
+    return None
+
+
+def cbor_make_credential_sign(
+    client_data_hash: bytes, cred: Fido2Credential, user_verification: bool
+) -> bytes:
+    privkey = cred.private_key()
+    pubkey = nist256p1.publickey(privkey, False)
+
+    flags = _AUTH_FLAG_UP | _AUTH_FLAG_AT
+    if user_verification:
+        flags |= _AUTH_FLAG_UV
+
+    # Encode the authenticator data (Credential ID, its public key and extensions).
+    credential_pub_key = cbor.encode(
+        {
+            _COSE_ALG_KEY: _COSE_ALG_ES256,
+            _COSE_KEY_TYPE_KEY: _COSE_KEY_TYPE_EC2,
+            _COSE_CURVE_KEY: _COSE_CURVE_P256,
+            _COSE_X_COORD_KEY: pubkey[1:33],
+            _COSE_Y_COORD_KEY: pubkey[33:],
+        }
+    )
+
+    att_cred_data = (
+        _AAGUID + len(cred.id).to_bytes(2, "big") + cred.id + credential_pub_key
+    )
+
+    extensions = b""
+    if cred.hmac_secret:
+        extensions = cbor.encode({"hmac-secret": True})
+        flags |= _AUTH_FLAG_ED
+
+    authenticator_data = (
+        cred.rp_id_hash
+        + bytes([flags])
+        + b"\x00\x00\x00\x00"
+        + att_cred_data
+        + extensions
+    )
+
+    # Compute self-attestation signature of the authenticator data.
+    dig = hashlib.sha256()
+    dig.update(authenticator_data)
+    dig.update(client_data_hash)
+    sig = nist256p1.sign(privkey, dig.digest(), False)
+    sig = der.encode_seq((sig[1:33], sig[33:]))
+
+    # Encode the authenticatorMakeCredential response data.
+    return cbor.encode(
+        {
+            _MAKECRED_RESP_FMT: "packed",
+            _MAKECRED_RESP_AUTH_DATA: authenticator_data,
+            _MAKECRED_RESP_ATT_STMT: {"alg": _COSE_ALG_ES256, "sig": sig},
+        }
+    )
+
+
+def cbor_get_assertion(req: Cmd, dialog_mgr: DialogManager) -> Optional[Cmd]:
+    if not storage.is_initialized():
+        if __debug__:
+            log.warning(__name__, "not initialized")
+        return cbor_error(req.cid, _ERR_OPERATION_DENIED)
+
+    try:
+        param = cbor.decode(req.data[1:])
+        rp_id = param[_GETASSERT_CMD_RP_ID]
+        rp_id_hash = hashlib.sha256(rp_id).digest()
+
```

Comment:
According to the Webauthn you should check 'if all the supplied parameters are syntactically well-formed and of the correct length' right here. CTAP2 doesn't say nothing about the position of the check.

Reply:
See above.